third_party/move/move-vm/types/src/value_serde.rs

```diff
@@ -0,0 +1,343 @@
+            // FIXME Add checks!
+            MoveTypeLayout::U64 => u64::deserialize(deserializer)?,
+            MoveTypeLayout::U128 => u128::deserialize(deserializer)? as u64,
+            layout if is_derived_string_struct_layout(layout) => {
+                // Here, we make sure we convert identifiers to fixed-size Move
+                // values. This is needed because we charge gas based on the resource
+                // size with identifiers inside, and so it has to be deterministic.
+                let value = DeserializationSeed {
+                    native_deserializer: None::<&NativeValueSimpleSerDe>,
+                    layout,
+                }
+                .deserialize(deserializer)?;
+                let (bytes, _width) = value
+                    .value_as::<Struct>()
+                    .map(derived_string_struct_to_bytes_and_length)
+                    .expect("TODO")
+                    .expect("TODO");
+                from_utf8_bytes::<u64>(bytes).expect("TODO")
+            },
+            _ => return Err(D::Error::custom("Failed serialization")),
+        };
+        let sized_id = SizedID::from(encoded);
+        Ok(Value::native_value(sized_id))
+    }
+}
+
+impl CustomSerialize for NativeValueSimpleSerDe {
+    fn custom_serialize<S: Serializer>(
+        &self,
+        serializer: S,
+        _tag: &IdentifierMappingKind,
+        layout: &MoveTypeLayout,
+        sized_id: SizedID,
+    ) -> Result<S::Ok, S::Error> {
+        match layout {
+            // FIXME Add checks!
+            MoveTypeLayout::U64 => serializer.serialize_u64(sized_id.into()),
+            MoveTypeLayout::U128 => {
+                let encoded: u64 = sized_id.into();
+                serializer.serialize_u128(encoded as u128)
+            },
+            layout if is_derived_string_struct_layout(layout) => {
+                let value = into_derived_string_struct(sized_id)
+                    .ok_or_else(|| S::Error::custom("Failed serialization"))?;
+                SerializationReadyValue {
+                    native_serializer: None::<&NativeValueSimpleSerDe>,
+                    layout,
+                    value: &value.0,
+                }
+                .serialize(serializer)
+            },
+            _ => Err(S::Error::custom("Failed serialization")),
+        }
+    }
+}
+
+pub fn deserialize_and_allow_native_values(bytes: &[u8], layout: &MoveTypeLayout) -> Option<Value> {
+    let native_deserializer = NativeValueSimpleSerDe;
+    let seed = DeserializationSeed {
+        native_deserializer: Some(&native_deserializer),
+        layout,
+    };
+    bcs::from_bytes_seed(seed, bytes).ok()
+}
+
+pub fn serialize_and_allow_native_values(
+    value: &Value,
+    layout: &MoveTypeLayout,
+) -> Option<Vec<u8>> {
+    let native_serializer = NativeValueSimpleSerDe;
+    let value = SerializationReadyValue {
+        native_serializer: Some(&native_serializer),
+        layout,
+        value: &value.0,
```

Comment:
What does value.0 do here?

Reply:
Here `value` is `Value` but serialization is defined for inner private type `ValueImpl`, so `&value.0` gives `&ValueImpl` and we can serialize it